NDArithmeticMixin and NDUncertainty Refactor

MSeifert04 opened the pull request:
Initial try of splitting the large #4234. More information will follow when the splitting is done.

**edit 12/22/2015: added checklist**

*Next steps:*

+ [ ]  Rework metadata arithmetic so that it does not enforce any specific methods of combining the metadata -- that works by calling methods on the metadata but that returns `None` with a warning if the methods don't exist. `_arithmetic_meta`
+ [ ] Make thin wrapper class around a meta if it doesn't have any arithmetic operations.

* Then, if that looks promising:*

+ [ ] Add wrapper classes for other properties.

MSeifert04 commented:
Short Summary
- ``UnknownUncertainty`` is now the default uncertainty for ``NDData``-like classes.
- ``NDUncertainty`` has now adopted the internal mechanics of ``StdDevUncertainty``.
- ``StdDevUncertainty`` now propagates uncertainties always correct and can use a given ``correlation``
- ``NDArithmeticMixin`` has classmethods that allow arithmetic operations where the ``first`` element is not a ``NDData``-like class but the result should be.
- ``NDArithmeticMixin`` has now different methods that operate on the attributes. Also different ways of handling the meta in arithmetic operations are now possible.

I think I forgot to mention some further changes and I hope this is not to much change for you to give me some comments.

MSeifert04 commented:
One issue I found about the ``Uncertainty`` refactor was how to organize the classes. Since ``NDUncertainty`` is a ``MetaClass`` but it does everything except the actual propagation. As an alternative I could make ``NDUncertainty`` like ``NDDataBase`` to a class that only implements an interface (``UnknownUncertainty`` would then be like a Minimal working implementation of it) and make the mechanics (like ``__init__`` and propagation abstractmethods) a Mixin with ``StdDevUncertainty`` being a working implementation of it. I hope you understand what I mean, if not I could wrap this idea in code (just superficially).

But maybe I'm thinking too much about it and it is fine just like it is.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+            function that is used by the name of the operation better than by
+            the function itself. Suppose we have ``data`` that is a
+            `~numpy.matrix` or ``Pandas.DataFrame`` we need to use different
+            UFUNCS (maybe) and that explodes the number of if/elif checks in
```
I think this (need to use different UFUNCS) isn't actually an issue per http://pandas.pydata.org/pandas-docs/stable/dsintro.html#dataframe-interoperability-with-numpy-functions

Even if it were an issue I think it would be better to make `operation` a function that defaults to the appropriate numpy ufunc because then the user can substitute a different, more appropriate, ufunc without needing to subclass or add a special case via a PR.

MSeifert04 replied:
Just to get this straight before I change it:
``add`` should provide the numpy ufunc (``np.add``) and pass it to ``_arithmetic`` as ``operation``? So that another ufunc could be provided by calling ``_arithmetic`` directly?
Another thing: How should I evaluate the function in ``_arithmetics_meta``: compare the function name or simply apply the ``operation``-parameter?
For ``uncertainty_propagation`` I'll need to compare the ``operation`` in a way, should I compare it like ``operation is np.add`` or by ``operation.__name__ == 'add'``?

I'm not sure about what would be best, that was one another reason why I thought string comparisons were the best way to go.

mwcraig replied:
> `add` should provide the numpy ufunc (`np.add`) and pass it to `_arithmetic` as operation? So that another ufunc could be provided by calling `_arithmetic` directly?

Mostly, yes. I think what I said before about substituting a different function without subclassing doesn't actually work, but a subclass that wanted to use a different addition could do so by just overriding `add` without needing to change anything in `_arithmetic`.

The other advantage of making this change is that you eliminate the need for the duplicate logic in `_arithmetic_data` and `_arithmetic_meta`

mwcraig replied:
> Another thing: How should I evaluate the function in `_arithmetics_meta`: compare the function name or simply apply the `operation`-parameter?

I would just apply the parameter.

mwcraig replied:
> For uncertainty_propagation I'll need to compare the operation in a way, should I compare it like operation is `np.add` or by `operation.__name__ == 'add'`?

Good question...looking over the code, I'd lean towards using `operation.__name__`. What about an approach like that in the current implementation [1]? That adds an argument to `_arithmetic` that just gets passed through to the uncertainty call but that maybe isn't so bad.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+                if isinstance(self.wcs, WCS) and isinstance(operand.wcs, WCS):
+                    log.info("Astropy.WCS does not allow equality checks.")
+                else:
+                    log.info("WCS is not equal.")
```
Given that the user can choose to compare WCS or not, I think an exception be raised here instead of warning if the two don't match.

MSeifert04 replied:
in both cases or just when it's not an astropy.wcs.WCS object?

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
-        return self._arithmetic(
-            operand, propagate_uncertainties, "subtraction", np.subtract)
+            # Both have WCS information. If requested compare the WCS and
+            # issue a warning if they are not equal
```
In the current version of `NDArithmeticMixin` there is a check `self.wcs != operand.wcs`. DO you know if this fails if the two WCS are the same?

wkerzendorf replied:
here similar to my comments above: overwrite `__ne__`.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
-            operand, propagate_uncertainties, "subtraction", np.subtract)
+            # Both have WCS information. If requested compare the WCS and
+            # issue a warning if they are not equal
+            # TODO: Currently (astropy 1.04) this is a bit tricky because no
```
Would astropy#4085 address this when fixed? *Not* suggesting that you try to fix that one...but if comparison does not work properly, and the issue is distinct from astropy#4085, could you please open a new issue with a short code snippet that will reproduce the issue? I gather the problem is that two identical WCS objects (identical in the sense that one is a `deepcopy` of the other) do not compare as the same?

MSeifert04 replied:
Yes, ``astropy.wcs.WCS`` does not implement a ``__eq__`` method so python defaults to check if the ``wcs1 is wcs2`` is ``True``. I'll supply an example later or tomorrow.

wkerzendorf replied:
yep - that's what I mean. It should for our case. Maybe make an super thin subclass can accomplish that (and later `gwcs` can implement that natively`)

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+                # that is a scalar... (ok, at least if we don't want to allow
+                # the meta to contain numpy arrays :-) )
+                if operand.data.size != 1:
+                    log.info("The second operand is not a scalar. Cannot "
```
I think this should raise an exception instead of logging an information message.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+                # Same thing but reversed
+                if self.data.size != 1:
+                    log.info("The first operand is not a scalar. Cannot"
+                             "operate meta keywords with arrays.")
```
Same comment as above.

mwcraig commented on astropy/nddata/mixins/tests/test_ndarithmetic.py:
```diff
@@ -6,463 +6,803 @@
+    (np.array(5) * u.Mpc, np.array(10)),
+    (np.array(5), np.array(10) * u.s),
+    ])
+def test_arithmetics_data_unit_NOT_identical(data1, data2):
```
nitpick: could you please lower-case the function name?

wkerzendorf replied:
PEP8 syntax requirement not nitpick :wink:

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+        else:
+            # Make sure these masks are boolean or ndarrays of bools so this
+            # kind of mask operation is possible.
+            if self.mask is not None:
```
I think the checks below are too restrictive and are unnecessary...why not simply do the mask operation `self.mask | operand.mask`? If it fails an exception will presumably be raised...if the desire is to catch the exception that is raised if the operation fails then it could be put inside a try/except.

MSeifert04 replied:
Unfortunatly the ``|`` operator works on most (string arrays raise an exception but it works on every numerical array) kinds of numpy arrays but the result is sometimes quite strange.

```
import numpy as np
a = np.array([1,0,1])
b = np.array([1,0,0])
a | b
# array([1, 0, 1], dtype=int32) this is what is happening to booleans too
b = np.array([1,0,2])
a | b
# array([1, 0, 3], dtype=int32) not what is expected...
```

mwcraig replied:
what about doing this: `np.array(a | b, dtype='bool')`?

Then I get a bool even with integer arrays:

```python
import numpy as np
a = np.array([1,0,1])
b = np.array([1,0,2])
np.array(a | b, dtype='bool')
# array([ True, False,  True], dtype=bool)
```

wkerzendorf replied:
This again shows that I believe the individual components of `NDData` need to decide if they can be added/subtracted/multiplied/divided. This would only mean that one moves a lot of the logic into light classes or the existing classes. another thing shouldn't it be `and` instead of `or`.

mwcraig commented:
@MSeifert04 -- lots of inline comments, I still want to try running the tests locally and see if it breaks anything in ccdproc :). Hoping to do that today.

MSeifert04 commented:
@mwcraig Thanks for the comments, I've adressed every obvious comment but left replies where I wasn't sure how to proceed.

If you need any help with ``ccdproc`` I can help if you want.

mwcraig commented:
@jehturner -- would you mind looking over this? It would impact `NDDataArithmetic`. While comments on the code itself are certainly welcome, it would be useful if you could try this out with what you are developing with `NDDataArithmetic` and report back whether it breaks anything.

jehturner commented:
OK. Any sane changes to the arithmetic are unlikely to break things for me, as I'm not systematically using that functionality yet (though obviously that's the intention). Most of the calculations are done by the external processes I'm initially wrapping and then gradually replacing, something I'm just completing the outline of. I can certainly do a quick check but it might take a few days as I'm juggling 2 deadlines. Anyway, thanks for asking and thanks to Michael for improving the code.

jehturner commented:
I confirm that this does not break my code (in a couple of dozen quick tests). Sorry it has taken so long to comment (has it really been 3 weeks?). I haven't yet looked closely at what these changes actually do but I'm not noticing any unexpected problems.

embray commented:
Thanks for checking it out @jehturner --the fact that it works for you is encouraging.

I hope to review these changes soon myself.

MSeifert04 commented:
@mwcraig @embray Should I incorporate @mwcraig suggestions or still wait a bit longer?

@jehturner Thanks for the tests.

mwcraig commented:
@MSeifert04 -- hold off for a couple more days, hoping to get comments form @wkerzendorf (he was traveling for most of November).

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
     When subclassing, be sure to list the superclasses in the correct order
```
Is there a way to automate or force the correct order? @embray you know a lot of python magic.

mhvk replied:
It is probably worth checking what makes it required to have a certain order. Often, one can write things such that order does not matter (but certainly not always).

mwcraig replied:
The MRO dictates the order - a helpful reference is http://nedbatchelder.com/blog/201210/multiple_inheritance_is_hard.html and the comments underneath.

If the mixin overrides a method in `NDDataBase` you want the mixin listed before the base class.

It might be better to simply point out that order matters and provide a reference....

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+    This class only aims at covering the most common cases so there are certain
+    restrictions on the saved attributes::
+
+        - ``uncertainty`` : has to be something that has a `NDUncertainty`-like
```
Correct me if I'm wrong - I believe that the idea was that it is an arbitrary object that has an attribute `.uncertainty_type`

MSeifert04 replied:
That was actually shortsighted of me, with the current implementation one only needed to pass the right ``propagate`` method for the uncertainty object. Probably I should switch back to the propagate-method parameter. But since there is no actual API what parameters should be passed to the ``propagate`` method makes it really hard to cover all cases.

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+
+    Notes
+    -----
+    1. It is not tried to decompose the units, mainly due to the internal
```
I wonder if we need to note this as this is the standard (and sensible behaviour) - see `km/s / Mpc` should not be `Hz` :wink:

MSeifert04 replied:
Well sometimes it makes sense, sometimes one would like to decompose the units. So I thought I should mention it somewhere.

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+        >>> from astropy.nddata import *
+        >>> class NDDataWithMath(NDArithmeticMixin, NDData): pass
+        >>> nd = NDDataWithMath([1,2,3], unit='meter')
+        >>> nd_inv = nd.ic_division(1, nd)
```
Why can't we just do `nd_inv = 1 / nd`?

MSeifert04 replied:
Because someone in APE said, that for ``NDData`` the magic mathematical methods should not be used. And without them it's getting really hard (I can't think of a way) to simulate ``__radd__``. If ``__add__`` for the first element isn't possible.

mwcraig replied:
IIRC an issue that comes up is if you try to add an `NDData` to a `Quantity`. Both override `__add__` and `__radd_` and the result ends up being of the type of whatever the first operand is. I think I was the one who raised this in the APE and it could be that I'm wrong...but overriding the arithmetic operators should be a separate PR (if it is ever done).

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+        # operations with numbers, lists, numpy arrays, numpy masked arrays
+        # astropy quantities, masked quantities and of other subclasses of
+        # NDData
+        operand = self.__class__(operand)
```
this is nice.

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+        # enforces that the dtype is NOT objects)?
+
+        kwargs = {}
+        # First check that the WCS allows the arithmetic operation
```
I would attach this to the `wcs` object. In the end, `NDData` just makes the operation `wcs1 + wcs2` and the either `wcs` object will complain if that doesn't work. you only need to check if both of them are `None` which is a special case.

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+        # quantity)
+        result = self._arithmetic_data(operation, operand, **kwds)
+        # Determine the other properties
+        if propagate_uncertainties is None:
```
There is an `NDUncertainty` class that in the end was thought to be a mixin. @mhvk how far along is that?

mhvk replied:
@wkerzendorf - My `Variable` class (#3715) is quite complete and does track uncertainties properly for all numpy `ufunc`. I've not worked on it further since I posted it, however, since there really should be a bit of general thought in when to stop tracking correlations.

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+        else:
+            kwargs['uncertainty'] = self._arithmetic_uncertainty(
+                operation, operand, result, uncertainty_correlation, **kwds)
+        if handle_mask is None:
```
Maybe this should be part of the `mask` as well.

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
-        if self.unit is None and operand.unit is None:
-            if result_unit is dimensionless_unscaled:
-                result_unit = None
+        if handle_meta is None:
```
I think similar to `wcs` this maybe should be part of the metaclass - you just do the arithmetic on the meta class and it will know what to do. This would then also be the same for tables, which should show the same behaviour.

MSeifert04 replied:
I never worked with metaclasses yet and what I've read you need to define the metaclass as superclass (at least I heard so for python 3.+). Also we allow everything that is a subclass from ``dict``, how would we get the metaclass in there without breaking something. Or do you mean a wrapper like ``NDUncertainty``?

Also I think that the container should define the way the properties are handled/propagated since the context might be different? Adding two dicts normally means something completly different from my implementation here.

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+            # Neither has uncertainties so the result should have none.
+            return None
         elif self.uncertainty is None:
-            result.uncertainty = operand_uncertainty
```
this definitely **was** wrong.

wkerzendorf commented:
@mwcraig am currently in proposal hell (which will freeze over on the 14th of December). I think this PR warrants a skype discussion as it is a major contribution to `NDData`. 

@MSeifert04 you are brave! This is definitely one of the most tricky things that astropy has to offer - so thank you for looking into this! As far as I can see almost all of the functionality is there - however I think structuring the arithmetic in the classes for `wcs`, `mask`, `meta` is a useful way to keep the whole thing modular and the code shorter and tidier.

mwcraig commented:
@wkerzendorf -- thanks for taking a look! Agreed this will be a skype pull request (especially because @MSeifert04 has also written #4271 and #4270). Classes end for me in a couple days, so I'll have time (finally) to pull together a meeting targeted for the week after your proposals are due.  I'll encourage people to take a look at these three PRs on the astropy-dev list next week.

mwcraig commented:
I took a look at the wrapper options @MSeifert04 has put together in a few gists and wonder if we should just go with something much simpler for now: 

Imagine this PR just adds the `_arithmetic_meta` method but that method defaults to doing nothing, or to putting, inside of a try/except block, something like: `meta1.operation(meta2)`?

Then if the meta implements the operation, good, if not, then that gets caught. It seems like the wrapper approach is an elaborate way of avoiding an exception...

Just adding `_arithemtic_meta` would still be an improvement over the current situation, since there would be only one method to override if one wanted to include metadata operations (as opposed to overriding the whole `_arithmetic` method). 

Partially I'm looking for ways to simplify this a bit....

hamogu commented:
I did not review this code in details, but on first sight it seems to me that it addresses a similar problem as #4459 does for metadata in tables. Could the arithmetic of `NDData` use the same registry of meta merging classes?

mwcraig commented:
> Imagine this PR just adds the _arithmetic_meta method but that method defaults to doing nothing, or to putting, inside of a try/except block, something like: meta1.operation(meta2)?

@MSeifert04 -- what do you think about this (from longer comment above)?

MSeifert04 commented:
@mwcraig I've thought about it and think you are right: Having Mixins or Wrapper for the classes are probably not really convenient.

But I would prefer giving the ``_arithmetic`` method a callable for ``operation``, ``meta``, ``mask`` and ``wcs`` and then call that function. For ``mask`` it would be an optional parameter with default ``numpy.bitwise_or`` for ``meta`` it could be a function in ``astropy.utils.metadata`` (maybe I will move the current meta there in the future or some adapted version of it) or simply ``None`` but with ``WCS`` I'm not quite sure or is there a python built-in for the ``==`` operator? operation will be filled as before with the numpy callables in the ``add``, ... methods.

That would require very little work from my side and I think will let users much freedom. Since I already pass the ``**kwargs`` to all the functions giving these functions parameters is also possible.

Do you think that's a reasonable option?

wkerzendorf commented:
The main issue, I have is that I think it is common to have the same metadata handling for different subclasses (i.e. `Spectrum1D` and `Image2D`). If we implement it on the class, I will always need to subclass and change it on the class itself, rather than just telling my meta-data what to do. but the given option is better than what is currently implemented.

MSeifert04 commented:
You are definitly right but at some point either the user must decide which meta-subclass he uses or it must be coded in some class.

For any use-case it probably doesn't matter at all since any arithmetic operation depends on the context. Which keywords should be modified or added or updated. In the end an addition (with meta) with a purpose on ``NDData`` will probably defined in a seperate function or method and will be setup so that the right action will be taken on the keywords. For example adding two images or spectra will take a different action for example on the exposure time than on the airmass and one needs to distinguish between ``NDData`` + ``NDData`` and ``NDData`` + value and use different calculations.

Probably there is a senseable way to do this with mixins/subclasses of meta but that's a bit beyond me. A function on the other hand can accomplish this quite easy.

mwcraig commented:
> Do you think that's a reasonable option?

@MSeifert04 -- yes, please go ahead with it. I wouldn't try anything fancy regarding metadata here given #4459. Sorry it took so long to reply, the email alert got buried.

mwcraig commented:
@embray -- did you still want to look this over?

MSeifert04 commented:
@mwcraig I've pushed the changes (I hope I got it right) but I do not understand why the tests are failing.

mwcraig commented:
@MSeifert04 -- I'm not sure what is up with the `fitting` import errors, but the `ConfigAlias` ones are because they were part of an old configuration system that was pulled out in #4446.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,141 @@
-from ..nduncertainty import IncompatibleUncertaintiesException
+from ..nduncertainty import NDUncertainty
+from ...units import dimensionless_unscaled
+from ...config import ConfigAlias
```
Drop this import...

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,141 @@
+
+WARN_UNSUPPORTED_CORRELATED = ConfigAlias(
+    '0.4', 'WARN_UNSUPPORTED_CORRELATED', 'warn_unsupported_correlated',
+    'astropy.nddata.nddata', 'astropy.nddata')
```
This can be removed completely. The actual configuration item is defined in `nddata/__init__.py`

MSeifert04 replied:
@mwcraig Thanks for the hint. I deleted these items. Since I made the possibility to give ``correlation`` as parameter I think this wasn't even used here anymore.

MSeifert04 commented:
@mwcraig Seems like the ``ConfigAlias`` also caused the ``fitting`` problems. Unfortunatly there are some Linux test fails. These happen because ``np.logical_or`` does not throw a ValueError when faced with ``string`` input. I'm not sure if I should just delete this part of the test or if I should make it work.

It shouldn't be too common that someone tries a ``string`` mask.

mwcraig commented:
@MSeifert04 -- I think it is fine to take that test out. The idea that the user is responsible for making sure the mask does something sensible has been built in since the beginning, so there is no need for an exhaustive set of tests of all possible bad masks. 

Once the tests are passing I'll mark this as ready for final review.

MSeifert04 commented:
@wkerzendorf @mwcraig You originally wanted to allow also for ``methods`` for doing the arithmetic on ``meta`` and ``wcs``. Would it be an option if I allowed for ``strings`` as ``compare_wcs``, ``handle_meta`` that would be interpreted as methods of ``meta`` or ``wcs``?

Unfortunatly the ``WCS().wcs.compare`` method (very useful in this case) is a method of an attribute of the ``wcs`` object and therefore it does not work with ``hasattr(self.wcs, 'wcs.compare')``. I'm not sure how one could bypass something like this without special casing ``WCS`` instances.

embray commented:
Can someone provide me a short walkthrough of this so I know where to begin reviewing this?

MSeifert04 commented:
I'm not sure if I did not forgot about something:

Short Summary:
- ``UnknownUncertainty`` is now the default uncertainty for ``NDData``-like classes. (already provisionally implemented in #4270)
- ``NDUncertainty`` has now adopted the internal mechanics of ``StdDevUncertainty``. (one rather general change is that the propagation now alway
- ``StdDevUncertainty`` now propagates uncertainties always correct and can use a given ``correlation``
- ``NDArithmeticMixin`` has classmethods that allow arithmetic operations where the ``first`` element is not a ``NDData``-like class but the result should be.
- ``NDArithmeticMixin`` has now different methods that operate on the attributes.  (``_arithmetics_*`` methods).
- Each arithmetic operation provides a ``func`` for adding the ``data`` and the user can define a function for handling the ``meta`` (default is None), ``mask`` (default is ``np.logical_or``) and compare the ``wcs`` (I wanted to wrap ``astropy.wcs.Wcsprm.compare`` inside a function and make it useable for this kind of comparison but I'll leave that for later)

I think I forgot to mention some further changes and I hope this is not to much change for you to give me some comments.

MSeifert04 commented:
oups, didn't want tp close it.

@embray:

There were a lot of changes, I'm not sure how to begin either. I've rewritten this class so often I did lose the overview.

``NDUncertainty``:
- Is adapted to the previous ``StdDevUncertainty`` (to handle ``unit``, ...)
- Every propagation calls ``NDUncertainty.propagate`` instead of the ``NDUncertainty.propagate_add``.
- ``NDUncertainty`` accepts a ``correlation`` parameter for propagation.

``StdDevUncertainty``
- Implements ``correlation`` and ``unit`` handling.

``UnknownUncertainty``
- Provisionally implemented in #4270 as wrapper for uncertainties that have no ``uncertainty_type``.

``NDArithmeticMixin``
- Implements ``classmethods`` for inverse arithmetic operations.
- ``add``, etc. provide the ``func`` that does the arithmetics on ``data`` with ``unit``
- user can specifiy which ``func`` should handle ``meta`` (default None), ``mask`` (default ``np.logical_or``) and a function to compare ``wcs`` (wanted later to wrap ``astropy.wcs.Wcsprm.compare`` for that but currently None).
- ``arithmetics`` of each attribute (except ``unit``) is now in a seperate method to easy subclassing in case one wants to implement something exceptional.
- ``arithmetics`` allows operations with everything that can be converted to a ``NDData`` object.

embray commented:
@MSeifert04 I know, it's hard to keep track :)  Thanks for the overview--this will help get started at least.  Hopefully you won't have to do too much more work on this :)

mwcraig commented:
@MSeifert04 -- I've blocked off some time on Thursday for giving this one more look too. Almost done, I think!

MSeifert04 commented:
I think the test failure is not because of me.

@mwcraig - Nice!

One general question: Since #4270 is merged should I clean up the stuff in this PR that isn't needed anymore?

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
 
 
+# TODO: General question of implementing an Operator class #3861? (embray)
+# Maybe a bit too big for these changes.
```
I think you can remove this TODO

mwcraig commented:
@MSeifert04 -- I'd hold off on the rebase until #4242 is in too.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+    from .utils import format_doc
+except ImportError:
+    # TODO: Delete this and rebase if #4242 is merged.
+    def format_doc(docstring, *args, **kwargs):
```
Looks like #4242 is close to done...so I wouldn't remove and rebase this *today* but once #4242 is merged go ahead.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+            return func
+        return set_docstring
+
+# TODO: Maybe not nice to pollute globals but I felt the same way about
```
Is this still a TODO? If not, convert to a plain comment about why you chose to put it in globals.

Also, since `__all__` is set above these globals won't affect the `nddata` namespace.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+
+    Parameters
+    ----------
+    operand : `NDData`-like instance or convertable to one.
```
convertable → convertible

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+
+    propagate_uncertainties : `bool` or ``None``, optional
+        If ``None`` the result will have no uncertainty. If ``False`` the
+        result will have a copied version of the first operand (or the
```
Maybe a minor rephrase:

If ``False`` the result will have a copied version of the first operand that has an uncertainty.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+    propagate_uncertainties : `bool` or ``None``, optional
+        If ``None`` the result will have no uncertainty. If ``False`` the
+        result will have a copied version of the first operand (or the
+        second if the first had no uncertainty). In case this is ``True``
```
"In case this is" → "If"

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+        result will have a copied version of the first operand (or the
+        second if the first had no uncertainty). In case this is ``True``
+        the result will have a correctly propagated uncertainty from the
+        uncertainties of the operands but this assumes that the uncertainties
```
Is it that it must be `StdDevUncertainty` or that the uncertainty must have a propagate method? If the latter then better to say that, I think.

MSeifert04 replied:
The documentation is wrong it only must be an ``NDUncertainty`` but for everything except ``StdDevUncertainty`` the ``propagate`` (or better the ``_convert_uncertainty``) raises an Error.

It is possible though to rewrite ``_arithmatics_uncertainty`` to either:
- check for ``hasattr(self.uncertainty, 'propagate')`` instead of ``isinstance(self.uncertainty, NDUncertainty)``. This makes the documentation really hard since I would need to explain: 
  1. What happens if one (and only one) of both uncertainties is None => it wraps it inside another ``other | self.uncertainty.__class__(None)`` 
  2. that the uncertainty implements a ``propagate`` method 
  3. that the propagate method must have the same signature as ``NDUncertainty.propagate`` 
  4. That the result-parameter for ``propagate`` may be an ``astropy.units.Quantity`` (actually this one isn't really necessary because it is mentioned in ``NDUncertainty_propagate``).
- Edit the documentation that it needs to be an ``NDUncertainty``-like class that implements propagation.

I'd lean to the second option and let subclasses that want to implement ``propagation`` for other uncertainties to override ``_arithmatics_uncertainty`` - With the splitting of methods this would be quite easy and probably only a one-liner. But that's just easier for me, I can see the advantages of the first option. :-)

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+
+    handle_mask : callable, ``False`` or ``None``, optional
+        If ``None`` the result will have no mask. If ``False`` the result
+        will have a copied version of the mask of the first operand (or
```
See comment about similar wording for `propagate_uncertainties`

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+        if the first operand has no mask then the one from the second is
+        taken). If it is a callable then the specified callable must
+        create the results ``mask`` and if necessary provide a copy.
+        Default is `numpy.logical_or`.
```
What happens if `handle_mask` is `True`?

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+        create the results ``mask`` and if necessary provide a copy.
+        Default is `numpy.logical_or`.
+
+    handle_meta : callable, ``False`` or ``None``, optional
```
Same comments as on `handle_mask`

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+        Default is ``False``.
+
+    compare_wcs : callable, ``False`` or ``None``, optional
+        If ``None`` the result will have no wcs and no comparison between
```
Again, same as `handle_mask`

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+        was given otherwise it raises a ``ValueError`` if the comparison was
+        not successful. Default is ``False``.
+
+    uncertainty_correlation : ``Number`` or `~numpy.ndarray`, optional
```
I would change the docstring to something like "Correlation between the uncertainties" then but the formula in the notes.

MSeifert04 replied:
Now that you mention it: Should I put the formula inside a ``.. math::``? And I just realized that I used the ``Covariance`` (dAB = dA \* dB \* cor) instead of the ``Correlation`` (cor = dAB / (dA * dB)) formula?

What about:
"Correlation between the uncertainties:
.. math:: 
    cor = d_{AB} / (d_A * d_B)"

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+
+    Notes
+    -----
+    In case a ``callable`` is used for ``mask``, ``wcs`` or ``meta`` the
```
"In case" → "If"

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+    parameters. If the callable also needs additional parameters these can be
+    defined as ``kwargs`` and must start with ``"wcs_"`` (for wcs callable) or
+    ``"meta_"`` (for meta callable). This startstring is removed before the
+    callable is called.
```
This works, but I'm wondering if there is a better way, e.g. one of the `kwargs` could be a `dict` with keys `wcs`, `meta`, etc, whose value is a dictionary of optional arguments for the callables. It seems a little magical for the leading string to be automatically stripped out.

mwcraig replied:
After reading the rest I'm inclined to make my suggestion a low-priority change. @embray may have a suggestion as to the right way to go here, but I'm inclined to get this merged even if we think another approach is better for passing in the arguments for the callables, then open a separate PR for that other approach.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,14 +8,138 @@
+
+    This method is avaiable as classmethod in order to allow arithmetic
+    operations between arbitary objects as long as they are convertable to
+    the class that called the method. Therefor the name prefix ``ic_`` which
```
"Therefor" → "Therefore"

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+    interface you can alter the ``propagate_uncertainties`` parameter in
+    :meth:`NDArithmeticMixin.add`. ``None`` means that the result will have no
+    uncertainty, ``False`` means it takes the uncertainty of the first operand
+    (if this does not exist from the second operand) as the results
```
"results" → "result's"

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+    :meth:`NDArithmeticMixin.add`. ``None`` means that the result will have no
+    uncertainty, ``False`` means it takes the uncertainty of the first operand
+    (if this does not exist from the second operand) as the results
+    uncertainty. This behaviour is also explained in the help-page for the
```
"help-page" → "docstring" ?

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+            unit) or as quantity if at least one had a unit.
+
+        kwargs : `dict`
+            the kwargs should contain all the other attributes (besides data
```
"the" → "The"

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+
+        kwargs : `dict`
+            the kwargs should contain all the other attributes (besides data
+            and unit) to create a new instance for the result. Creating the
```
"to" → "needed to"

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+            new instance is up to the calling method, for example
+            :meth:`NDArithmeticMixin.add`.
+
+            TODO: Reword this? (embray)
```
Can be removed now, I think.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+        Base method which calculates the result of the arithmetic operation.
+
+        This class determines the result of the arithmetic operation on the
+        ``data`` with respect to their units and then forwards to other methods
```
Do you mean "this method" instead of "this class"? Also, I think "with respect to" → "including"

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+
+        # Find the appropriate keywords for the appropriate method (not sure
+        # if data and uncertainty are ever used ...)
+        # TODO: Ask about data and uncertainty :-)
```
remove TODO

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+        kwargs = {}
+
+        # First check that the WCS allows the arithmetic operation
+        if compare_wcs is None:
```
I think there is a problem with the logic in this case and the following few that are similar. The `else` is reached whenever `compare_wcs` is neither `None` nor `False`. We don't really need to worry about all cases (e.g. if someone passes in `compare_wcs=5`) there are two values we might reasonably expect users to pass in:

+ a callable (no problem with that)
+ `True` -- since `False` is meaningful one might reasonably expect that `True` is also meaningful. In fact, `True` gets you to the `else` clause and an error is raised in `_arithmetic_wcs` because `True` isn't callable.

Two thoughts for changing this:

+ add an `elif callable(compare_wcs)` that contains the call to `_arithmetic_wcs` and in the `else` raise a `ValueError`.
+ Change the allowed values for `compare_wcs` to be `None` or the string `"first"` or a callable. 

At the moment I'd lean towards the second option because it would make the intent of the behavior of what is currently done when the value is `False` much more clear.

MSeifert04 replied:
Actually this is just a question about expectations of the end-user. I remember my frustration with ``matplotlib.pyplot.imshow(interpolation=..)``. That's ugly as hell as there are two values for ``interpolation``: ``None`` and ``'none'``, which do completly different things. So I'm very glad you pointed it out. As the programmer I thought I made it quite clear what is allowed: ``None``, ``False`` and ``callable`` and what happens. But I'm quite close to the code and I see your point about the expectation that if ``False`` is allowed then ``True`` should/could be allowed. But what about the expectation in case of ``'first'`` wouldn't someone then assume that it could also be ``"second"``? 

I think this is something that should be discussed/decided before I change anything. :-)

mwcraig replied:
I like your suggestion of `first_found` -- that is clear and unambiguous. I've also hit the `None` vs `none` thing in matplotlib *shudder*...

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+        Returns
+        -------
+        result_data : `~numpy.ndarray` or `~astropy.units.Quantity`
+            if both operands had no unit the resulting data is a simple numpy
```
"if" → "If"

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+        result : `~astropy.units.Quantity` or `~numpy.ndarray`
+            The result of :meth:`NDArithmeticMixin._arithmetic_data`.
+
+        correlation : `Number` or `~numpy.ndarray`
```
Remove single backticks around "Number" and lowercase it.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+            return handle_mask(self.mask, operand.mask, **kwds)
 
-        return result
+    def _arithmetic_wcs(self, operation, operand, compare_wcs, **kwds):
```
Does `operation` need to be an argument here? I realize that keeps the signature similar to the other `_arithmetic_` methods.

If you do keep it, please change the the documentation for for `operation` in this method to indicate that it is ignored.

MSeifert04 replied:
Yes, I included it simply to have the same method signature for wcs, mask and meta. I included that it is ignored.

mwcraig replied:
Sounds good, thanks!

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -3,17 +3,71 @@
+try:
+    from ..utils import format_doc
+except ImportError:
+    # TODO: Delete this and rebase if #4242 is merged.
```
Just another reminder to change this once #4242 is merged.

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -3,17 +3,71 @@
+    The data for the uncertainty of b in a {operator} b
+result_data : `~numpy.ndarray` instance or `~astropy.units.Quantity`
+    The data array that is the result of the {operation}.
+correlation: `Number` or `~numpy.ndarray`
```
Remove backticks around "Number" and lowercase it

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -3,17 +3,71 @@
+    The data array that is the result of the {operation}.
+correlation: `Number` or `~numpy.ndarray`
+    Array or scalar representing the correlation. If the subclass does not
+    support correlated uncertainties this will be replaced by 0 (uncorrelated).
```
Haven't read the code below yet, so ignore if the code addresses this, but it seems dangerous to silently change correlated uncertainties to uncorrelated ones.

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -3,17 +3,71 @@
+Notes
+-----
+Handling units (especially if the differ from the parents unit) is done in here
+but be aware that since there are no checks for the unit that incorrect
```
"unit that" → "unit, "

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -30,155 +84,187 @@ class MissingDataAssociationException(Exception):
+       directly.
+
+    2. NDUncertainty takes a ``unit`` parameter and if the `array` is something
+       with a unit (like a `~astropy.units.Quantity` of another
```
"of" → "or" ?

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -30,155 +84,187 @@ class MissingDataAssociationException(Exception):
+
+    2. NDUncertainty takes a ``unit`` parameter and if the `array` is something
+       with a unit (like a `~astropy.units.Quantity` of another
+       `NDUncertainty`) the `unit` parameter is considered the True unit of
```
"True" → "true"

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -30,155 +84,187 @@ class MissingDataAssociationException(Exception):
+            # Check if two units are given and take the explicit one then.
+            if (unit is not None and array.unit is not None and
+                    unit != array.unit):
+                # TODO : Clarify it (see NDData.init for same problem)?
```
Is this still a todo?

MSeifert04 replied:
You silently accepted it in #4270 (I think there is still the TODO). That was if given two units (one from data, one explicitly) should raise an error or raise a warning.

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -190,191 +276,483 @@ def parent_nddata(self):
+            The result of the arithmetic operations. This saves some duplicate
+            calculations.
+
+        correlation: ``Number`` or `~numpy.ndarray`
```
No backticks around "Number" and lowercase it.

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -190,191 +276,483 @@ def parent_nddata(self):
+        # Check if the subclass supports correlation
+        if not self.supports_correlated:
+            if correlation != 0:
+                log.info("{0} does not support uncertainty propagation with "
```
Why not raise an error here instead? Logs are pretty easily ignored/overlooked....

MSeifert04 replied:
Changed it and documentation.

MSeifert04 replied:
Just a note here: The line above makes real trouble. I have written the whole correlation support only for numbers and given an ``numpy.ndarray`` all the comparisons with 0 fail because ``ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()``

I think for now I'll change correlation to only support ``numbers`` and add ``np.ndarray`` correlations in a seperate PR. This could take some time, expecially writing tests.

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -190,191 +276,483 @@ def parent_nddata(self):
+    uncertainty has a unit that differs from (but is convertable to) the
+    parents `NDData` unit but converts the unit of the propagated uncertainty
+    to the unit of the resulting data. Also support for correlation is possible
+    but that requires that the correlation is an input it cannot handle
```
"input it" → "input. It"

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -190,191 +276,483 @@ def parent_nddata(self):
+
+            if self.unit != other_uncert.unit:
+                # In case the two uncertainties (or data) have different units
+                # we need to use quantity operations. The case where only on
```
"on" → "one"

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -190,191 +276,483 @@ def parent_nddata(self):
-             + (other_nddata.uncertainty.array/other_nddata.data)**2) *
-             result_data)
+        else:
+            # Formula sigma = |AB|*sqrt((dA/A)**2+(dB/B)**2-2*dA/A*dB/B*cor)
```
I think `|AB|` → `|A/B|`?

I should also add the comment with the error propagation formulae are VERY helpful!

mwcraig commented:
Once again, thanks for this contribution. There are a number of line edits, mostly grammar.

A couple of more substantive (but not too time-consuming) changes, also noted above inline:

+ I think it would be clearer if the options for `compare_wcs`, `handle_mask`, etc, were `None`, `"first"` or a callable instead of `None`, `False`, or a callable (with `True` resulting in a error when it is treated as a callable).

+ If during error propagation a correlation is given but `supports_correlated` is `False`, an exception should be raised instead of setting the correlation to zero and logging it. 

As noted above, let's not try changing how you strip out the `wcs_`, etc., from the keywords for the callables unless someone has really strong feelings about it. It would be easy to change the behavior later and I'd like to get this merged soon!

mwcraig commented:
> I think the test failure is not because of me.

yep, restarted the failing test, the underlying issues have been fixed in master.

MSeifert04 commented:
@mwcraig - Thanks for the revision. I changed all the documentation things except the ones i answered inline. Because some of the documentation changes are nearby to one of your open issues I havn't pushed the change yet in case they got outdated.

I wasn't to sure about the ``False`` for meta/mask/wcs but ``'first'`` has the problem that one might assume that ``'second'`` is also possible? And with ``'first'`` I wouldn't assume that it takes the second if the first was ``None``. Maybe ``'first found'``? (see also https://github.com/astropy/astropy/pull/4272#discussion_r50529571)

Your second point is correct, since ``correlation`` is by default 0 one should raise an ``Error`` if the user specified a different one.

The ``kwargs`` split was provisional. I just thought that giving named parameters to ``add`` would be better than defining ``dict`` before calling the method. But for example there is the same parameter name in the function that is called for ``wcs`` and for ``meta`` some conflicts may arise. I guess there are probably better solutions using ``python.inspect`` (or something else) but for now the solution works (even if it is not very nice). So I like the solution to solve it in a different PR. :-)

https://github.com/astropy/astropy/pull/4272#discussion_r50478768 is something that's actually still a ``TODO`` even though I forgot to include it in https://github.com/astropy/astropy/blob/master/astropy/nddata/nddata.py#L235. I don't know if/where we talked about the problem of giving an explicit and implicit unit to ``NDData.__init__``.

MSeifert04 commented:
@mwcraig Since I have found a serious error when using correlation with as ``np.ndarray`` I decided to fix it and push. I'll include here a list of *open* comments in case they get outdated.

- https://github.com/astropy/astropy/pull/4272#discussion_r50461510
- https://github.com/astropy/astropy/pull/4272#discussion_r50463209
- https://github.com/astropy/astropy/pull/4272#discussion_r50463863
- https://github.com/astropy/astropy/pull/4272#discussion_r50473240
- https://github.com/astropy/astropy/pull/4272#discussion_r50476495
- https://github.com/astropy/astropy/pull/4272#discussion_r50478768
- https://github.com/astropy/astropy/pull/4272#discussion_r50479653

If any of those is outdated and I haven't adressed the issue properly please feel free to drop a line-note again. Or should I do it?

hamogu commented:
This is the best discussed PR with the most detailed review I have ever seen.
Thank you so much @MSeifert04 for writing it and @mwcraig for reviewing it in this detail. I'm absolutely looking forward to use the stuff in my classes.

mwcraig commented:
:blush: Thanks, @hamogu!

mwcraig commented:
@MSeifert04 -- here is what I'm thinking in terms of merge strategy, since this PR has some overlap with a few others:

Monday (or Tuesday at latest) I'll merge #4242 and #4503 (which implements the fix for WCS comparison that will be in 1.0.9 and 1.1.2). After that you will need to rebase (and #4503 will cause a conflict that is hopefully easy to fix) and then this, at long last, gets merged in!

MSeifert04 commented:
@hamogu - Thank you.

@mwcraig - Ok, as soon as #4242 (and #4503) is merged I'll rebase and change the ``False`` to ``'first_found'`` (https://github.com/astropy/astropy/pull/4272#discussion_r50629450). Thanks for all your help and the time for reviewing and talking about this PR :thumbsup: :-)

jehturner commented:
@mwcraig wrote:
> It would be particularly nice to get one more set of eyes on the part of the PR 
> that modifies StdDevUncertainty and handles correlated error...so feel free to 
> look at that even if you don't have time to review the entire PR.

Sorry if I'm late in commenting on this (especially after Matt pointed me to it some time ago; it's just hard to find enough time). Support for more general arithmetic sounds helpful, as does working error propagation :smiley:. Yea.

I have not been thinking about correlated errors in my recent work but I'm just having a quick look at this FWIW and wondering about some very high-level things. Sorry if these are silly questions...

I would expect to deal with the sort of correlated errors I run into (between neighbouring pixels) by propagating sparse covariance matrices, if necessary (I suppose in place of the simpler uncertainty arrays). I think here you are more concerned with correlations between different arrays, such that `uncertainty_correlation` and `correlation` are simpler matrices with the same dimensionality as the data, right? I think a bit more explanation of what those parameters are supposed to be might be helpful. Just as a side note, the former name seems to suggest that it's just the uncertainties that are correlated but presumably the data values themselves are likewise correlated, so it could just be `correlation` without the qualifier?

It initially struck me as odd that the correlation would be supplied as an argument to the arithmetic operation rather than being a class attribute, the same as the existing error `array`; after all, it seems like a characteristic of the *data* rather than the process. However, as I say, I'm used to dealing with correlations *within* a dataset rather than between them -- I'm not sure how else you *could* do this in general without some overarching data structure to relate the different NDData instances (yikes). How does the correlation get propagated by your application, if not as part of the NDData instances? I suppose it's all a bit too abstract without knowing your use case.

Maybe I can play with this a bit later, but don't let me hold anything up...

mhvk commented:
Correlated errors is more than a little tricky; see #3715 for an implementation. I'd suggest to discuss it separately.

mwcraig commented:
@MSeifert04 -- I've merged the other two pull requests that affected this one (or maybe didn't....since this is still showing no conflicts?).

I'll merge later today.

I've created a separate issue for adding a bit of documentation about the correlated errors (#4538), and agree with @mhvk that we should continue the discussion of @jehturner's points in a separate issue or pull request.

mwcraig commented:
Ah, @MSeifert04, guess you do need to rebase this.

jehturner commented:
OK, thanks. I was mostly responding to the request to look at this stuff and am not currently worrying much about correlated errors (though it does come up with resampled IFU data). I'm curious, though, whether this seems potentially compatible with what you're aiming towards, @mhvk? Looks like you would probably deprecate NDUncertainty altogether?

MSeifert04 commented:
@jehturner - Sorry, I haven't seen your comments. I just used the ``uncertainty_correlation`` as given in the error propagation formulas (I don't recall which book I had them from but they are also given here: https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulas). This is just a shortcut if one really tries to use some *known* correlation.

@mwcraig - I rebased and deleted the provisional code parts. I hope I have time the next days to make an accurate Changelog entry.

MSeifert04 commented:
I've added a small changelog entry. I think it covers everything related to public methods.

eteq commented:
@MSeifert04 @mwcraig and all the reviewers - Lots of great work here, +:100: !

I have one concern, though, and that's the lack of (narrative) documentation and examples, particularly with the correlation stuff (which is notoriously tricky).  I see that #4538 was created as a reminder to do that.  *But*, in my experience, if something gets merged with a separate "do the documentation" issue, often the documentation doesn't get written until it's far too late (i.e., someone mis-used it or gave up on it because they didn't understand it).

That said, I understand you might want to get it in soon for other changes.  So @mwcraig, I'm fine with you making a command decision about whether my concern here is warranted or if you're confident you and/or @MSeifert04 will get the docs in for this in a reasonably timely fashion before the key points have escaped your heads.

eteq commented:
Oh one other thing that's more informational than anything else: I think we'll also want to try to at least get a start on something with `Quantity` and uncertainties for v1.2 - maybe @mhvk's #3715, or maybe something simpler that then #3715 ends up building on.  We will certainly want to try to figure out how to harmonize that with `nddata`'s uncertainties... But I think it's probably better to let them operate seperately at least initially so that we have APIs that make sense for those objects, and then later on try to hook them together better.

mhvk commented:
Agreed to let `Quantity` and `NDData` proceed separately; my notes here were really about making people aware of #3715, since it does correlations perfectly (though the resulting derivative matrices can get rather out of hand -- which is what needs thought).

mwcraig commented:
@eteq -- I'm going to go ahead and merge this; @MSeifert04 has an outline in an ipython notebook in #4538. 

Agreed it would have been better to flesh out the docs at the same time, but this has lingered long enough I'd like to get it closed.  

I'll work with @MSeifert04 to divide up the work on the docs with a target date of a week or two.
